src/modules/update/common/update-common.c

```diff
@@ -0,0 +1,488 @@
+        SOL_WRN("Invalid response code from [%s] when checking for update: %d",
+            handle->url, http_response->response_code);
+        handle->cb_check((void *)handle->user_data, -http_response->response_code, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    if (!streq(http_response->content_type, "application/json")) {
+        SOL_WRN("Invalid content type of response: [%s] expected [application/json]",
+            http_response->content_type);
+        handle->cb_check((void *)handle->user_data, -EINVAL, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    if (!http_response->content.used) {
+        SOL_WRN("Empty response for version check");
+        handle->cb_check((void *)handle->user_data, -EINVAL, NULL);
+        delete_handle(handle);
+        return;
+    }
+
+    sol_json_scanner_init(&scanner, http_response->content.data, http_response->content.used);
+    SOL_JSON_SCANNER_OBJECT_LOOP (&scanner, &token, &key, &value, reason) {
+        if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "url"))
+            response.url = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "version"))
+            response.version = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "hash"))
+            response.hash = sol_json_token_get_unescaped_string_copy(&value);
+        else if (SOL_JSON_TOKEN_STR_LITERAL_EQ(&key, "hash-algorithm"))
+            response.hash_algorithm = sol_json_token_get_unescaped_string_copy(&value);
+        else
+            SOL_WRN("Unknown response member: %.*s",
+                SOL_STR_SLICE_PRINT(sol_json_token_to_slice(&token)));
+    }
+
+    /* If some member is empty, give up */
+    if (!(response.url && response.version
+        && response.hash && response.hash_algorithm)) {
+        SOL_WRN("Malformed check response");
+        handle->cb_check((void *)handle->user_data, -EINVAL, NULL);
+        free_response_members(&response);
+        delete_handle(handle);
+        return;
+    }
+
+    handle->cb_check((void *)handle->user_data, 0, &response);
+    free_response_members(&response);
+    delete_handle(handle);
+
+    return;
+}
+
+static void
+on_digest_ready_cb(void *data, struct sol_message_digest *md, struct sol_blob *output)
+{
+    struct sol_update_handle *handle = data;
+    struct sol_buffer buffer = SOL_BUFFER_INIT_EMPTY;
+    struct sol_str_slice slice = sol_str_slice_from_blob(output);
+    int r = 0;
+
+    r = sol_buffer_append_as_base16(&buffer, slice, false);
+    SOL_INT_CHECK_GOTO(r, < 0, end);
+
+    if (buffer.used != strlen(handle->hash) ||
+        strncmp(buffer.data, handle->hash, buffer.used != 0)) {
+
+        r = -EINVAL;
+        SOL_WRN("Expected hash differs of download file hash, expected [%s], found [%.*s]",
+            handle->hash, (int)buffer.used, (char *)buffer.data);
+    }
+
+end:
+    handle->cb_hash(handle, r);
```

Comment:
same as with check, flag that we're calling the cb so we don't double free. We can do something like store `handle->cb_hash` to a stack variable, set `handle->cb_hash = NULL` and check that condition inside `_cancel()`